apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/query/ReadBridgeService.java

```diff
@@ -89,15 +93,24 @@ public Multi<RawDocument> searchDocuments(
 
       // load the candidates
       Multi<RawDocument> candidates =
-          documentsResolver
-              .getDocuments(queryExecutor, keyspace, collection, paginator)
-
-              // limit to requested page size only to stop fetching extra docs
-              .select()
-              .first(paginator.docPageSize);
+          documentsResolver.getDocuments(queryExecutor, keyspace, collection, paginator);
 
       // then populate
-      return populateCandidates(candidates, keyspace, collection, nestedPopulate(context));
+      return populateCandidates(candidates, keyspace, collection, nestedPopulate(context))
+
+          // ensure that the document matches the expression
+          // this is to avoid race condition that a document can change between
+          // candidate searching and population
+          .filter(
+              document -> {
+                Map<String, EvalRule<FilterExpression>> rules = EvalEngine.booleanRules();
+                rules.put(FilterExpression.EXPR_TYPE, new RawDocumentEvalRule(document));
+                return EvalEngine.evaluate(expression, rules);
+              })
+
+          // limit to requested page size only to stop fetching extra docs
+          .select()
+          .first(paginator.docPageSize);
```

Comment:
we already select the first paginator.docPageSize on line 97, what is the second for ?

Reply:
The first on line `97` was moved here.. Not sure I actually understand, do I have two `.first(paginator.docPageSize)` calls now?

Reply:
My bad, I  missed the line removals.